### ISSUE TITLE
`omega` gives values that are not within [-1,1] bounds

### ISSUE THREAD

**Issue by mamonu** [networkx/networkx#5064]
According to the paper this metric has been based upon, [found here](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3604768/) , omega needs to be between [-1,1].

However nx.omega gives as a result many values that are not within [-1,1] bounds

### Current Behavior

values not being within the expected [-1,1] bounds

### Expected Behavior

values should be within [-1,1] bounds

### Steps to Reproduce

```python
import networkx as nx
G = nx.barbell_graph(5,1)
print(nx.omega(G,seed=1)) # result is -1.505815972222222
```
### Environment

Python version: 3.7.4
NetworkX version: 2.5.1

**Comment by rossbar** [comment]
Thanks for the report - I can reproduce this on the development branch (96831f99). I also tried the barbell graph example you've provided with different seeds and noticed that the result was always < -1.

**Comment by jtrim-ons** [comment]
I've only had a very quick look at the paper you linked to @mamonu, but one question that springs to mind is whether the [-1,1] interval is just something the authors observed for graphs they looked at rather than something that is true for all graphs. Certainly the wording "Moreover, values of [omega] are restricted to the interval -1 to 1 regardless of network size." makes it sound like a general thing, but the definition of omega makes me wonder. Is there a proof in the paper that I've missed?

**Comment by jtrim-ons** [comment]
You could also see if you get close to some of the paper's values for the karate graph (`nx.karate()`) from the first row of Table 1.

**Comment by jtrim-ons** [comment]
... I got 0.35 for the karate graph, whereas the paper says 0.08. But it might just be down to random variation.

**Comment by jtrim-ons** [comment]
I've had a little look at `algorithms/smallworld.py`, and noticed the following things that may be worth checking. (Unfortunately I don't have time to check them properly, and will have to stop looking any further at this issue.)

- C is calculated using transitivity in NetworkX but using the clustering coefficient in the paper. These can be very different (http://pages.stat.wisc.edu/~karlrohe/netsci/MeasuringTrianglesInGraphs.pdf)
- The `niter` and `nrand` parameters perhaps don't match the paper? I haven't checked this properly.
- I also noticed that the NetworkX implementation of *sigma* uses the `random_reference` function. It might be worth checking whether this unusual random graph generator is the standard way to calculate sigma.

**Comment by mamonu** [comment]
thanks for looking both. 

btw @jtrim-ons I used to work at ONS some years ago :)

its a `smallworld` indeed. (ba-ddum-tch)

**Comment by fabriz-io** [comment]
I would like to look into this issue including the comments from the discussion. From the contributor guide I could not identify if issues get assigned, so I would just start? If I am missing something, please let me know.

**Comment by fabriz-io** [comment]
I tried to keep it as short as possible, but I think the explanation is necessary for further discussion.

Following the comments from @jtrim-ons, I started by inspecting if the \[-1, 1\] bounds are meant to be tight. From what I understand,  yes, they should be tight bounds, even if they mention a deviation for small networks:

*Note that as the size of the network is reduced, the tails of the x curve deviate from the bounds of the [1,1] interval, particularly from 1. This finding is due to the fact that in a very small network the real path length never becomes very large relative to the path length of the random network.*

But from my understanding this part means that the values for omega do not converge (from within the bounds) to the bounds of the interval, regardless of the used rewiring probability.

Then, further following the comments from @jtrim-ons, I changed from transitivity to clustering coefficient (nx.average_clustering(G)) as used in the paper and also adjusted the values for niter and nrand (more details on this below). But both changes did not change the value for `omega` being around -1.5 for the graph mentioned by @mamonu (nx.barbell_graph(5, 1)).

Then I turned back to the paper and tried to understand in more detail the construction of the random and lattice reference and I noticed a clear difference. While the random reference graph seems to be constructed with the Maslov and Sneppen algorithm (I checked the values for some graphs against the examples in the paper and they match exactly), the lattice reference _"(...) was generated by using a modified version of the 'latticization' algorithm (Sporns and Zwi, 2004) found in the brain connectivity toolbox (www.brainconnectivity-toolbox.net) (Rubinov and Sporns, 2010) (...)"_. I checked the Matlab Code from this toolbox against the Sporns and Zwi implementation for the lattice reference and this seems to be fine. The key difference is the *adjusted* version they mention, which is explained in more detail in the supplemental material. In short: They do not calculate the mean clustering coefficient over several generated lattice reference graphs (as the omega implementation does now), but rather try to maximize the clustering coefficient  by only accepting a new lattice reference, if the clustering is higher. Otherwise the old lattice is kept. Changing this yields an omega of around -0.28 for the mentioned barbell graph (I tried different seeds and iterations and this is rather robust). With this adjustments I get for the Karate graph an omega of around -0.065 and for the word adjacency graph 0.55. (It should be noted that the N of the networks from the provided sources differ from the N of the provided examples, e.g. the word adjacency graph has 425 nodes in a single connected component and not 112)

If you agree with my findings, I would continue to implement the following:

- Changing from transitivity to clustering coefficient
- Changing default values for niter and nrand:
	- Note: Not all of this values are explicitly mentioned.
	- niter_lattice=5 is mentioned in the supplemental material
	- niter_random=10 and nrand_random=50 is mentioned in the section "Random and lattice network construction"
	- nrand_lattice is not explicitly mentioned and only described as "performed over several user-defined repetitions". For my calculations I used also 50, but I would question this choice as the default value, as e.g. the calculation of omega took 23 minutes for the word adjaceny graph with 425 nodes. Furthermore, for larger networks (N>1000) they completely switch to a "sliding-window approach" for the lattice generation, which is further described in the supplemental material. It could be worth digging into this for the sake of efficiency.
- Changing the calculation of clustering coefficient for the lattice as describe above
